### DISCUSSION TITLE
[SOLVED] How can I assign a labels trough Article filters

### DISCUSSION THREAD

**Discussion by BobSquarePants** [martinrotter/rssguard#772] [Q&A] [answered]
Hi everyone,

I've tried by myself to make a **Article filters** that will assign a label when the `title` or `content` match a keyword, but definitely JavaScript make not sens at all for me..

I'm sure it will be something close to -->

```javascript
var whitelist = [
  'test subject', 'blabla'
];
function filterMessage() {
  if (whitelist.some(i => msg.title.indexOf(i) != -1)) {
  msg.assignLabel(findLabelId('MyLabel'))
  } 
}
```

But I can't figure it out. thanks

**Comment by M1chael** [comment]
I did almost the same thing and see that you missed `msg.` part before `findLabelId`. But in my case it also doesn't work. No errors and console.log debugging confirms all my assumptions, but no labels assigned.

**Reply by M1chael** [reply]
Complete function I used:
```javascript
function filterMessage() {
  if (title_whitelist.some(i => msg.title.toUpperCase().indexOf(i.toUpperCase()) != -1)) {
    if (title_blacklist.some(i => msg.title.toUpperCase().indexOf(i.toUpperCase()) != -1) || contents_blacklist.some(i => msg.contents.toUpperCase().indexOf(i.toUpperCase()) != -1)) {
      return MessageObject.Ignore;
    } else {
      if (kernel.some(i => msg.contents.toUpperCase().indexOf(i.toUpperCase()) != -1)) {
         msg.assignLabel(msg.findLabelId('Linux kernel CVE'));
      }
      return MessageObject.Accept;
    }
  } else {
    return MessageObject.Ignore;
  }
}
```

**Reply by martinrotter** [reply]
it sounds about right, let me check, ALSO, pls post your full filtering script + sample feed URL on which to test

**Reply by M1chael** [reply]
[Feed sample](https://nvd.nist.gov/feeds/xml/cve/misc/nvd-rss-analyzed.xml).

Full filtering script:
```javascript
var kernel = [
  'Linux kernel'
];

var year = '-' + new Date().getFullYear() + '-';

var contents_blacklist = [
  'WordPress', 'Discourse', 'MediaWiki', 'IBM', 'Samsung', 'D-link', 'Mikrotik', 'Advancecomp', 'php', 'scada', 'Tenda', 'Dell', 'PLC', 'BlueZ', 'Bluetooth', 'cms', 'firmware', '** REJECT ** '
];

var title_blacklist = [
  'firmware', 'android', 'windows', 'azure', 'cms', 'stock_management_system', 'teleport', 'doctor-appointment-system'
];

var title_whitelist = [
  year
];

function filterMessage() {
  if (title_whitelist.some(i => msg.title.toUpperCase().indexOf(i.toUpperCase()) != -1)) {
    if (title_blacklist.some(i => msg.title.toUpperCase().indexOf(i.toUpperCase()) != -1) || contents_blacklist.some(i => msg.contents.toUpperCase().indexOf(i.toUpperCase()) != -1)) {
      return MessageObject.Ignore;
    } else {
      if (kernel.some(i => msg.contents.toUpperCase().indexOf(i.toUpperCase()) != -1)) {
         msg.assignLabel(msg.findLabelId('Linux kernel CVE'));
      }
      return MessageObject.Accept;
    }
  } else {
    return MessageObject.Ignore;
  }
}
```

Example of article I expect to be labeled:
```XML
<item rdf:about="https://web.nvd.nist.gov/view/vuln/detail?vulnId=CVE-2022-1651">
<title>CVE-2022-1651 (linux_kernel)</title>
<link>https://web.nvd.nist.gov/view/vuln/detail?vulnId=CVE-2022-1651</link>
<description>A memory leak flaw was found in the Linux kernel in acrn_dev_ioctl in the drivers/virt/acrn/hsm.c function in how the ACRN Device Model emulates virtual NICs in VM. This flaw allows a local privileged attacker to leak unauthorized kernel information, causing a denial of service.</description>
<dc:date>2022-07-26T17:15:08Z</dc:date>
</item>
```

But it doesn't:
![Screenshot from 2022-09-05 10-57-47e](https://user-images.githubusercontent.com/920901/188370773-c1e5812d-eab8-4b80-9a55-18fb88ebde9e.png)

Maybe it would be useful to add testing such things as labeling to the Script output tab in creating filters window.

**Reply by martinrotter** [reply]
I checked your situation and found a potential bug in the application which prevents your script from working. Your script is correct. I am trying to make it reasonably work.

Will post "semi-fix" with workaround soon + full fix later.

**Reply by martinrotter** [reply]
![image](https://user-images.githubusercontent.com/1255302/188387365-ea54a766-7807-4c30-b3c2-009aeaf43a6d.png)

OK, I added a semi-fix to allow it to work with "workaround" (see below), f0bb361e6.

Workaround is to use latest RSS Guard with the above fix (see development builds) and place line:

```js
msg.customId = msg.url;
```

to be the first line inside `filterMessage()` function. Like this (note I also changed name of the label for my testing):

```js
var kernel = [
  'Linux kernel'
];

var year = '-' + new Date().getFullYear() + '-';

var contents_blacklist = [
  'WordPress', 'Discourse', 'MediaWiki', 'IBM', 'Samsung', 'D-link', 'Mikrotik', 'Advancecomp', 'php', 'scada', 'Tenda', 'Dell', 'PLC', 'BlueZ', 'Bluetooth', 'cms', 'firmware', '** REJECT ** '
];

var title_blacklist = [
  'firmware', 'android', 'windows', 'azure', 'cms', 'stock_management_system', 'teleport', 'doctor-appointment-system'
];

var title_whitelist = [
  year
];

function filterMessage() {
  msg.customId = msg.url;

  if (title_whitelist.some(i => msg.title.toUpperCase().indexOf(i.toUpperCase()) != -1)) {
    if (title_blacklist.some(i => msg.title.toUpperCase().indexOf(i.toUpperCase()) != -1) || contents_blacklist.some(i => msg.contents.toUpperCase().indexOf(i.toUpperCase()) != -1)) {
      return MessageObject.Ignore;
    } else {
      if (kernel.some(i => msg.contents.toUpperCase().indexOf(i.toUpperCase()) != -1)) {
         msg.assignLabel(msg.findLabelId('Test test'));
      }
      return MessageObject.Accept;
    }
  } else {
    return MessageObject.Ignore;
  }
}
```

This is only temporary workaround but will also work after I push my full fix. Also note that this workaround will only reliably work for feeds whose entries DO HAVE UNIQUE URLs, which in your case seems to be true.

**Reply by M1chael** [reply]
Wow! Extremely fast fixing! You are incredible! Thanks a lot!

I switched to the AppImage from development builds and will report back when another liable to be labeled article comes to my feed.

**Reply by M1chael** [reply]
Yeah! Seems to be worked. Little minor bug though still there: after fetching articles RSSGuard shown 25 Labels, but only 10 of them were named. After clicking on the Labels or the group name digits became equal. Nothing critical, just FYI. Thanks again for fast fix!
![Untitled](https://user-images.githubusercontent.com/920901/188412337-b9fb6a6d-fc8a-4f25-b5db-1b20fd5a1b30.png)

**Reply by martinrotter** [reply]
Yes, that seems like a bug. Will check.

EDIT: Checked it and can reproduce. It is result of "THE" bug we workarounded above and will be likely fixed when I manage to fix the base problem here.

To outline what the problem is: we have kind of chicken-egg problem:
1. RSS Guard downloads articles.
2. Your article filter run and if your article filter assigns labels, **they are saved to database now** (this is the problematic behavior because the messages itself are not saved in DB so far).
3. Articles are saved to database.

**Comment by martinrotter** [comment, marked as answer]
> Hi everyone,
> 
> I've tried by myself to make a **Article filters** that will assign a label when the `title` or `content` match a keyword, but definitely JavaScript make not sens at all for me..
> 
> I'm sure it will be something close to -->
> 
> ```js
> var whitelist = [
>   'test subject', 'blabla'
> ];
> function filterMessage() {
>   if (whitelist.some(i => msg.title.indexOf(i) != -1)) {
>   msg.assignLabel(findLabelId('MyLabel'))
>   } 
> }
> ```
> 
> But I can't figure it out. thanks

Your script is missing fundamental part, `return` statement indicating whether message (modified) is actually accepted by the filter or not.

Fixed:
```js
var whitelist = [
  'qt', 'abc'
];
function filterMessage() {
  if (whitelist.some(i => msg.title.indexOf(i) != -1)) {
    console.log("inside");
    msg.assignLabel(msg.findLabelId('Test Test'))
  }

return MessageObject.Accept;
}
```

**Reply by BobSquarePants** [reply]
Thank you @martinrotter ! it worked !!!

of topic, would it possible to create a plugin for RSSguard (or a fork) that support Python for the filter instead of javascript ?

**Reply by martinrotter** [reply]
yes, it is possible, but i will not do it, someone else can do it of course, rssguard is open-source free software

**Reply by BobSquarePants** [reply]
Indeed, and thanks to have created RSSguard as a FOSS software :heart: 
Maybe I'll do it in the future then :)
